### PR TITLE
Migrate metric test from Boost to Catch2Migration

### PR DESCRIPTION
**Pull request by Abilityguy**
Referencing issue #2523 

Replaced 
```
BOOST_REQUIRE_CLOSE_FRACTION(x, y, 1e-z) with 
```
with
```
REQUIRE(x == Approx(y).epsilon(1e-(z+2)));
```

### CODE REVIEW

File: src/mlpack/tests/metric_test.cpp [hunk 1]
```diff
@@ -92,50 +90,54 @@ BOOST_AUTO_TEST_CASE(LINFMetricTest)
...
+          Approx(0.612479577).epsilon(1e-6));
 
   // Use hieght - width representation of bounding boxes.
   // Bounding boxes represent {x0, y0, h, w}.
   bbox1 << 49 << 75 << 154 << 50;
   bbox2 << 42 << 78 << 144 << 48;
   // Value calculated using Python interpreter.
-  BOOST_REQUIRE_CLOSE(IoU<>::Evaluate(bbox1, bbox2), 0.7898879, 1e-4);
+  REQUIRE(IoU<>::Evaluate(bbox1, bbox2) ==
+          Approx(0.7898879).epsilon(1e-6));
```
Comment (rcurtin) [inline src/mlpack/tests/metric_test.cpp:131]: I think this (and other lines) can fit in 80 characters, so no need to wrap them. :+1:

**Review by rcurtin — COMMENTED** [review]
Hey @Abilityguy, thanks for working on this.  Everything looks good to me except there is a small style issue that we should fix. :+1:

The test failure appears to be something that's fixed by #2642.

**Comment by Abilityguy** [comment]
Thanks @rcurtin for the review. I missed the style fixes. Will take care next time.

**Comment by kartikdutt18** [comment]
I have resolved the conflict, Let's wait for a green build. Thanks for the work.

**Comment by kartikdutt18** [comment]
Hey @Abilityguy, Could you pull again locally and remove `mean_shift_test.cpp` from mlpack_tests to fix the build. I accidentally left it there and is causing the build failure. Thanks.

**Comment by Abilityguy** [comment]
Sure @kartikdutt18. Made the fix.

**Comment by jeffin143** [comment]
@Abilityguy 
The following change is correct 

```
BOOST_REQUIRE_CLOSE_FRACTION(x, y, 1e-z) with 
with

REQUIRE(x == Approx(y).epsilon(1e-(z+2)));
```

**Comment by Abilityguy** [comment]
Cool. Thanks @jeffin143.

**Comment by jeffin143** [comment]
> Cool. Thanks @jeffin143.

Sorry @Abilityguy 
Just went through the docs again
Actually 
`BOOST_REQUIRE_CLOSE(x, 1.23, 0.15) --> BOOST_REQUIRE_CLOSE_FRACTION(x, 1.23, 0.0015)`

`BOOST_REQUIRE_CLOSE(x, 1.23, 0.15) -> REQUIRE(result == Approx(1.23).epsilon(0.0015))`

and there for By equating both of these 

`BOOST_REQUIRE_CLOSE_FRACTION(x, 1.23, 0.0015) -->  REQUIRE(result == Approx(1.23).epsilon(0.0015))`

Which means just change the line and no values change

File: src/mlpack/tests/metric_test.cpp [hunk 2]
```diff
@@ -330,34 +330,32 @@ BOOST_AUTO_TEST_CASE(BLEUScoreTest)
...
+  REQUIRE(bleu.TranslationLength() == 12);
+  REQUIRE(bleu.ReferenceLength() == 12);
 
   expectedPrecision = {0.692308f, 0.6f, 0.428571f, 0.25f};
   for (size_t i = 0; i < bleu.Precisions().size(); ++i)
   {
-    BOOST_REQUIRE_CLOSE_FRACTION(bleu.Precisions()[i],
-        expectedPrecision[i], 1e-04);
+    REQUIRE(bleu.Precisions()[i] ==
+            Approx(expectedPrecision[i]).epsilon(1e-5));
```
Comment (jeffin143) [inline src/mlpack/tests/metric_test.cpp:359]:
```suggestion
            Approx(expectedPrecision[i]).epsilon(1e-4));
```

File: src/mlpack/tests/metric_test.cpp [hunk 3]
```diff
@@ -330,34 +330,32 @@ BOOST_AUTO_TEST_CASE(BLEUScoreTest)
...
   }
 
   //! We will use smoothing function here by setting smooth to true.
   bleu.Evaluate(referenceCorpus, translationCorpus, true);
-  BOOST_REQUIRE_CLOSE_FRACTION(bleu.BLEUScore(), 0.459307, 1e-05);
-  BOOST_REQUIRE_EQUAL(bleu.BrevityPenalty(), 1.0);
-  BOOST_REQUIRE_EQUAL(bleu.Ratio(), 1.0);
-  BOOST_REQUIRE_EQUAL(bleu.TranslationLength(), 12);
-  BOOST_REQUIRE_EQUAL(bleu.ReferenceLength(), 12);
+  REQUIRE(bleu.BLEUScore() == Approx(0.459307).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/metric_test.cpp:349]:
```suggestion
  REQUIRE(bleu.BLEUScore() == Approx(0.459307).epsilon(1e-5));
```

File: src/mlpack/tests/metric_test.cpp [hunk 4]
```diff
@@ -330,34 +330,32 @@ BOOST_AUTO_TEST_CASE(BLEUScoreTest)
 
   //! We are not using smoothing function here.
   bleu.Evaluate(referenceCorpus, translationCorpus);
-  BOOST_REQUIRE_CLOSE_FRACTION(bleu.BLEUScore(), 0.0, 1e-05);
-  BOOST_REQUIRE_EQUAL(bleu.BrevityPenalty(), 1.0);
-  BOOST_REQUIRE_EQUAL(bleu.Ratio(), 1.0);
-  BOOST_REQUIRE_EQUAL(bleu.TranslationLength(), 12);
-  BOOST_REQUIRE_EQUAL(bleu.ReferenceLength(), 12);
+  REQUIRE(bleu.BLEUScore() == Approx(0.0).epsilon(1e-7));
```
Comment (jeffin143) [inline src/mlpack/tests/metric_test.cpp:333]:
```suggestion
  REQUIRE(bleu.BLEUScore() == Approx(0.0).epsilon(1e-5));
```

**Review by jeffin143 — COMMENTED** [review]
Last few bits

**Comment by Abilityguy** [comment]
Hey @jeffin143 , thanks for the suggestions and review. 
I made your suggested changes on my local system and all the test cases passed. 
I have made a commit reflecting the required changes.
Thanks again for the help.

**Comment by kartikdutt18** [comment]
Hey @Abilityguy, The failure seems unrelated, I have requested a re-run. Once it comes back green, I'll take another look. Thanks for this PR.

**Comment by Abilityguy** [comment]
Sure @kartikdutt18.

**Review by kartikdutt18 — APPROVED** [review]
Looks good to me. The build failure is unrelated. I'll take a closer look at this. Thanks for the contribution.

File: src/mlpack/tests/metric_test.cpp [hunk 5]
```diff
@@ -40,17 +38,17 @@ BOOST_AUTO_TEST_CASE(L1MetricTest)
 
   ManhattanDistance lMetric;
 
-  BOOST_REQUIRE_CLOSE((double) arma::accu(arma::abs(a1 - b1)),
-                      lMetric.Evaluate(a1, b1), 1e-5);
+  REQUIRE((double) arma::accu(arma::abs(a1 - b1)) ==
+          Approx(lMetric.Evaluate(a1, b1)).epsilon(1e-7));
```
Comment (zoq) [inline src/mlpack/tests/metric_test.cpp:42]:
```suggestion
      Approx(lMetric.Evaluate(a1, b1)).epsilon(1e-7));
```

See https://github.com/mlpack/mlpack/wiki/DesignGuidelines#tabbing for more information.

File: src/mlpack/tests/metric_test.cpp [hunk 6]
```diff
@@ -40,17 +38,17 @@ BOOST_AUTO_TEST_CASE(L1MetricTest)
...
 
-  BOOST_REQUIRE_CLOSE((double) arma::accu(arma::abs(a2 - b2)),
-                      lMetric.Evaluate(a2, b2), 1e-5);
+  REQUIRE((double) arma::accu(arma::abs(a2 - b2)) ==
+          Approx(lMetric.Evaluate(a2, b2)).epsilon(1e-7));
```
Comment (zoq) [inline src/mlpack/tests/metric_test.cpp:45]:
```suggestion
      Approx(lMetric.Evaluate(a2, b2)).epsilon(1e-7));
```

See https://github.com/mlpack/mlpack/wiki/DesignGuidelines#tabbing for more information. There are other places where we should apply the patch as well.

**Comment by Abilityguy** [comment]
Thanks for the help @zoq.
I have made the fixes.

**Review by birm — APPROVED** [review]

**Pull request merged**